libtracepoint-control-cpp/src/TracepointSession.cpp

```diff
@@ -74,107 +74,144 @@ RoundUpBufferSize(uint32_t pageSize, size_t bufferSize) noexcept
     return BufferSizeMax;
 }
 
-/*
-Manages a scatter-gather list of chunks of memory that need to be written to
-the perf.data file. This allows us to reduce the number of kernel calls.
-Instead of calling write() once per event (or twice if the event wraps), we
-call writev once for every 16 noncontinguous blocks of memory to be written.
-*/
-class IovecList
-{
-    static constexpr unsigned Max = 16;
-    unsigned m_used = 0;
-    iovec m_vecs[Max];
-
-public:
-
-    unsigned
-    RoomLeft() const noexcept
-    {
-        return Max - m_used;
-    }
+static tracepoint_decode::PerfEventSessionInfo
+MakeSessionInfo(uint32_t clockid) noexcept
+{
+    static const auto Billion = 1000000000u;
 
-    void
-    Add(uint8_t const* p, size_t c) noexcept
+    tracepoint_decode::PerfEventSessionInfo sessionInfo;
+    sessionInfo.SetClockid(clockid);
+
+    timespec monotonic;
+    timespec realtime;
+    if (0 == clock_gettime(clockid, &monotonic) &&
+        0 == clock_gettime(CLOCK_REALTIME, &realtime))
     {
-        assert(m_used < Max); // Caller is responsible for checking RoomLeft().
-        auto const lastVec = m_used - 1;
-        if (0 != m_used &&
-            p == static_cast<uint8_t*>(m_vecs[lastVec].iov_base) + m_vecs[lastVec].iov_len)
+        uint64_t monotonicNS, realtimeNS;
+        if (monotonic.tv_sec < realtime.tv_sec ||
+            (monotonic.tv_sec == realtime.tv_sec && monotonic.tv_nsec < realtime.tv_nsec))
         {
-            // This block immediately follows the last block. Merge the blocks.
-            m_vecs[lastVec].iov_len += c;
+            monotonicNS = 0;
+            realtimeNS = static_cast<uint64_t>(realtime.tv_sec - monotonic.tv_sec) * Billion
+                + realtime.tv_nsec - monotonic.tv_nsec;
         }
         else
         {
-            m_vecs[m_used].iov_base = const_cast<uint8_t*>(p);
-            m_vecs[m_used].iov_len = c;
-            m_used += 1;
+            realtimeNS = 0;
+            monotonicNS = static_cast<uint64_t>(monotonic.tv_sec - realtime.tv_sec) * Billion
+                + monotonic.tv_nsec - realtime.tv_nsec;
         }
+
+        sessionInfo.SetClockData(clockid, realtimeNS, monotonicNS);
     }
 
-    _Success_(return == 0) int
-    Flush(PerfDataFileWriter& output) noexcept
+    return sessionInfo;
+}
+
+namespace
+{
```

Comment:
Note: Add anonymous namespace so that these private helper classes are kept private.